build-system/routes/a4a-envelopes.js

```diff
@@ -91,40 +72,44 @@ app.use('/inabox-(friendly|safeframe)', (req, res) => {
 // http://localhost:8000/a4a[-3p]/proxy/s/www.washingtonpost.com/amphtml/news/post-politics/wp/2016/02/21/bernie-sanders-says-lower-turnout-contributed-to-his-nevada-loss-to-hillary-clinton/
 app.use('/a4a(|-3p)/', (req, res) => {
   const force3p = req.baseUrl.startsWith('/a4a-3p');
-  let adUrl = req.url;
   const templatePath = '/build-system/server-a4a-template.html';
-  const urlPrefix = getUrlPrefix(req);
-  if (!adUrl.startsWith('/proxy') && urlPrefix.includes('//localhost')) {
-    // This is a special case for testing. `localhost` URLs are transformed to
-    // `ads.localhost` to ensure that the iframe is fully x-origin.
-    adUrl = urlPrefix.replace('localhost', 'ads.localhost') + adUrl;
-  }
-  adUrl = addQueryParam(adUrl, 'inabox', 1);
+  const url = getInaboxUrl(req);
   fs.readFileAsync(process.cwd() + templatePath, 'utf8').then(template => {
-    const content = fillTemplate(template, adUrl, req.query)
+    const content = fillTemplate(template, url.href, req.query)
       .replace(/CHECKSIG/g, force3p || '')
       .replace(/DISABLE3PFALLBACK/g, !force3p);
     res.end(replaceUrls(SERVE_MODE, content));
   });
 });
 
-function getUrlPrefix(req) {
-  return req.protocol + '://' + req.headers.host;
-}
-
 /**
- * @param {string} url
- * @param {string} param
- * @param {*} value
- * @return {string}
+ * @param {Request} req
+ * @param {string|undefined} extraExperiments
+ * @return {!URL}
  */
-function addQueryParam(url, param, value) {
-  const paramValue =
-    encodeURIComponent(param) + '=' + encodeURIComponent(value);
-  if (!url.includes('?')) {
-    url += '?' + paramValue;
-  } else {
-    url += '&' + paramValue;
+function getInaboxUrl(req, extraExperiments) {
```

Comment:
Can we change extraExperiments to an array of experiment names and then `join(",")`? Too easy to mess up the comma.

Reply:
discussed offline. renamed `extraExperiments` .to `extraExperiment`